onnxruntime/python/tools/transformers/gpt2_beamsearch_tester.py

```diff
@@ -0,0 +1,435 @@
+# Copyright (c) Microsoft Corporation.  All rights reserved.
+# Licensed under the MIT License.  See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+# This script helps evaluation of GPT-2 model.
+import os
+import logging
+import torch
+import random
+import numpy
+import time
+import timeit
+import math
+import statistics
+from pathlib import Path
+from gpt2_tester import Gpt2Tester, Gpt2Metric
+from gpt2_beamsearch_helper import Gpt2BeamSearchHelper, Gpt2BeamSearchInputs
+from benchmark_helper import Precision
+
+logger = logging.getLogger(__name__)
+
+class Gpt2TesterFactory:
+    @staticmethod
+    def create_tester(tester_type="default"):
+        testers = {
+            "default": Gpt2Tester,
+            "beam_search_step": Gpt2BeamSearchTester,
+        }
+        w = testers[tester_type]
+        return w
+
+class Gpt2BeamSearchTester(Gpt2Tester):
+    def __init__(self,
+                 input_ids,
+                 position_ids,
+                 attention_mask,
+                 beam_select_idx,
+                 input_log_probs,
+                 input_unfinished_sents,
+                 prev_step_results,
+                 prev_step_scores,
+                 num_attention_heads,
+                 hidden_size,
+                 num_layer,
+                 beam_size,
+                 device,
+                 is_fp16=False,
+                 top_k=20,
+                 top_k_required_order=False,
+    ):
+        super().__init__(
+            input_ids,
+            position_ids,
+            attention_mask,
+            num_attention_heads,
+            hidden_size,
+            num_layer,
+            device,
+            is_fp16,
+            top_k,
+            top_k_required_order
+        )
+        self.input_length = input_ids.shape[-1]
+        self.n_layer = num_layer
+        self.beam_size = beam_size
+
+        self.beam_select_idx = beam_select_idx
+
+        self.input_log_probs = input_log_probs
+        self.input_unfinished_sents = input_unfinished_sents
+
+        self.prev_step_results = prev_step_results
+        self.prev_step_scores = prev_step_scores
+
+        self.last_state = None
+
+    def get_inputs(self) -> Gpt2BeamSearchInputs:
+        return Gpt2BeamSearchInputs(
+            self.input_ids,
+            self.position_ids,
+            self.attention_mask,
+            self.past,
+            self.beam_select_idx,
+            self.input_log_probs,
+            self.input_unfinished_sents,
+            self.prev_step_results,
+            self.prev_step_scores,
+        )
+
+    def update(self, output, step, device):
+        """
+        Update the inputs for next inference.
+        """
+        self.last_state = (
+            torch.from_numpy(output[0])
+            if isinstance(output[0], numpy.ndarray)
+            else output[0].clone().detach().cpu()
```

Comment:
Here the output is copied to CPU. Did you run test in GPU?

Reply:
Our one step beam search scenario is on CPU only. We didn't test it on GPU specifically.

I saw it passed GPU CI pipeline check. I can do a test on GPU at our side.

Reply:
I have pushed a few changes and made test. Now one step beam search works on GPU with fp16 model